sw/device/silicon_creator/lib/drivers/uart.c

```diff
@@ -9,101 +9,90 @@
 
 #include "sw/device/lib/arch/device.h"
 #include "sw/device/lib/base/bitfield.h"
-#include "sw/device/lib/base/mmio.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
 #include "sw/device/silicon_creator/lib/error.h"
 
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
 #include "uart_regs.h"  // Generated.
 
-#define NCO_WIDTH 16
-_Static_assert((1UL << NCO_WIDTH) - 1 == UART_CTRL_NCO_MASK,
-               "Bad value for NCO_WIDTH");
-
-static void uart_reset(const uart_t *uart) {
-  mmio_region_write32(uart->base_addr, UART_CTRL_REG_OFFSET, 0u);
+static void uart_reset(void) {
+  abs_mmio_write32(TOP_EARLGREY_UART0_BASE_ADDR + UART_CTRL_REG_OFFSET, 0u);
 
   // Write to the relevant bits clears the FIFOs.
   uint32_t reg = 0;
   reg = bitfield_bit32_write(reg, UART_FIFO_CTRL_RXRST_BIT, true);
   reg = bitfield_bit32_write(reg, UART_FIFO_CTRL_TXRST_BIT, true);
-  mmio_region_write32(uart->base_addr, UART_FIFO_CTRL_REG_OFFSET, reg);
-
-  mmio_region_write32(uart->base_addr, UART_OVRD_REG_OFFSET, 0u);
-  mmio_region_write32(uart->base_addr, UART_TIMEOUT_CTRL_REG_OFFSET, 0u);
-  mmio_region_write32(uart->base_addr, UART_INTR_ENABLE_REG_OFFSET, 0u);
-  mmio_region_write32(uart->base_addr, UART_INTR_STATE_REG_OFFSET, UINT32_MAX);
+  abs_mmio_write32(TOP_EARLGREY_UART0_BASE_ADDR + UART_FIFO_CTRL_REG_OFFSET,
```

Comment:
Seems like we need an offset-to-address function. In previous projects we had some such, members of a kind of " address map region" generated from the CSR objects.

Reply:
The notion of base regions and offsets has been removed from the Mask ROM.  The Mask ROM conceptualizes MMIO space generically as part of the 32-bit CPU address space.  This decision was made primarily as a size concern for secure validation of expectations between boot stages.  IOW, there are a number of MMIO reads and writes for which we remember the read or written values (in a scrambled form) so we can check whether or not the values still hold later in the boot process (or if there has been an attack/glitch).  Keeping both a `base` and `offset` around complicates the storage and scrambling scheme.

Overall, I prefer the base/offset scheme, but its not appropriate for the Mask ROM case.